enhancement(vector source, vector sink): batch events per rpc call.

JeanMertz opened the pull request:
This PR updates the Vector source/sink combo to batch events before sending requests from the sink to the source.

JeanMertz committed:
update vector sink to batch requests

Signed-off-by: Jean Mertz <git@jeanmertz.com>

JeanMertz committed:
fixup! update vector sink to batch requests

Signed-off-by: Jean Mertz <git@jeanmertz.com>

JeanMertz committed:
update vector source to accept batched events

Signed-off-by: Jean Mertz <git@jeanmertz.com>

JeanMertz commented on src/sources/vector/v2.rs:
```diff
@@ -42,24 +43,28 @@ impl proto::Service for Service {
+                .await
+                .map_err(|err| Status::unavailable(err.to_string()))?;
+
+            let status = match receiver {
+                Some(receiver) => receiver.await,
+                None => BatchStatus::Delivered,
+            };
+
+            match status {
+                BatchStatus::Errored => return Err(Status::internal("Delivery error")),
+                BatchStatus::Failed => return Err(Status::data_loss("Delivery failed")),
+                BatchStatus::Delivered => (),
+            }
```
@bruceg what do you propose should happen here if messages 1 and 2 or 5 are delivered, but 3 of 5 fails to be delivered. Should it stop processing 4 and 5, or try to process all events, and return some kind of status response that identifies which events failed to be delivered?

bruceg replied:
There should be a single `BatchNotifier` created for the whole array of events, which would produce a single status for the entire set. Move creating the notifier (and receiver) up out of the `map` around line 38, and then you should be able to `send_all` the entire batch at once and collect a single status at the end.

JeanMertz committed:
rework batched e2e acknowledgement

Signed-off-by: Jean Mertz <git@jeanmertz.com>

bruceg commented on proto/vector.proto:
```diff
@@ -3,11 +3,11 @@ package vector;
 
-message EventRequest {
-  event.EventWrapper message = 1;
+message PushEventsRequest {
+  repeated event.EventWrapper events = 1;
```
Now that 0.14 has gone out with the gRPC vector protocol, does this need a workaround to avoid breaking backwards compatibility?

JeanMertz replied:
That's a great point. I hadn't realized we already officially announced the new gRPC protocol.

Although having said that, both the [source](https://vector.dev/docs/reference/configuration/sources/vector/) and [sink](https://vector.dev/docs/reference/configuration/sinks/vector/) still default to v1 (e.g. the old TCP-based implementation), and neither document the `v2` switch to use gRPC, so maybe we're in the clear to make breaking changes here?

It'd at least be somewhat annoying to have to support both use-cases for something we only just shipped, and arent' actively using/promoting yet.

cc @binarylogic

spencergilbert replied:
We haven't "officially" announced it yet, but it is available

jszwedko replied:
It does appear to be undocumented so I think it's ok to break compatibility here. We should just avoid suggesting to anyone that they use it until it is documented (and thus officially released).

JeanMertz replied:
I'll go ahead and resolve this discussion. We'll ship this breaking change, without having to mark is a breaking, since we haven't promoted the `v2` version of this source/sink-combo yet, and the documentation doesn't mention it either.

People that have found this themselves without any docs are likely technical enough to figure out what has changed.

bruceg reviewed (commented):
LGTM except for the question about backwards compatibility

JeanMertz committed:
update tests

It turns out the encoded message does _not_ have newlines embedded, so
the comment explaining why 7 bytes needed to be trimmed seems to be
inaccurate.

After fixing the logic by only trimming the first 5 bytes, decoding
worked as expected.

I suspect the original implementation was actually trimming the
`PushEventsRequest` wrapper by trimming two more bytes, which then
allowed the embedded data to be decoded into `EventWrapper`.

In this case, we need to decode into `PushEventsRequest`, because we
have an array of events, and `prost::Message` cannot be implemented for
a foreign type, so we can't decode `Vec<EventWrapper>`, but _can_ decode
`PushEventsRequest`.

Thanks to @ktff for debugging this with me.

Signed-off-by: Jean Mertz <git@jeanmertz.com>

StephenWakely commented on proto/vector.proto:
```diff
@@ -3,11 +3,11 @@ package vector;
 
-message EventRequest {
-  event.EventWrapper message = 1;
+message PushEventsRequest {
```
What is the reason for renaming this to `PushEvents`? I'm not sure I understand the name..

StephenWakely replied:
Is it 1 Vector 'pushing' events to another?

JeanMertz replied:
Mostly consistency. This is used as the input type for the `PushEvents` service call. We also have a `HealthCheck` call that accepts `HealthCheckRequest`.

Additionally, given that we now push more than one event per request, renaming it from `Event` to `Events` made sense to me.

StephenWakely reviewed (approved):
Beautiful!

ktff reviewed (approved):
Nice
